### ISSUE TITLE
有关仿射变换的一些问题

### ISSUE THREAD

**Issue by huxiaopang666** [Ahuer-Lei/ADRNet#2]
作者您好，我想咨询一下您有关仿射变换校准的问题，您在实验中提到，您训练使用的数据对是先对两幅精确配准的大图进行随机仿射，然后裁剪相同区域生成不配准的图像对，这样处理的图像正常来说只带很少的黑色背景，但是在看到您结果展示的时候，感觉好像不是裁剪生成的结果直接做的。我按照您实验中的裁剪做法生成了图像对，但是进行逆仿射真值矩阵变换后，这种裁剪的moving fig是不能完全恢复到原来的空间位置去的。还望作者解释一下，或是劳烦您开源一下您的代码和使用的数据学习一下

**Comment by Ahuer-Lei** [comment]
裁剪或者不裁剪，对结果没有影响， 你是怎么算逆仿射真值矩阵的？， 代码后续会公开，最近较忙。

**Comment by huxiaopang666** [comment]
作者您好，我的图像处理过程是这样的，暂时不考虑异源数据，就是对单一张原始图像处理，首先我是随机生成一个2×3的仿射矩阵，然后是通过F.affine_grid和F.grid_sample去生成仿射变换后的图像，然后在它们的相同区域上进行裁剪，生成如下图所示的情况
![image](https://github.com/Ahuer-Lei/ADRNet/assets/125127115/a81da3cc-7113-46ff-8490-3195b9ac8661)
对刚才随机生成的仿射矩阵我在下面加上[0,0,1]后，因为其满秩，所以采用torch.inverse求取逆仿射矩阵，再将生成的逆仿射矩阵通过F.affine_grid和F.grid_sample应用到裁剪的moving img上，最后虽然能恢复空间信息，但是还是空间上不精确对齐，存在一定的像素偏移，结果如下图所示
![image](https://github.com/Ahuer-Lei/ADRNet/assets/125127115/a47f9e59-6a49-4104-958f-c1304670159b)
![image](https://github.com/Ahuer-Lei/ADRNet/assets/125127115/1feab0eb-135d-44ab-ab5f-2ed1e5eeb110)
（请看恢复出来的图像的左下角，确实空间信息不能精确对准）
还请麻烦作者帮我分析下问题所在

**Comment by Ahuer-Lei** [comment]
随机生成一个2×3的仿射参数，是怎么生成的？

**Comment by huxiaopang666** [comment]
作者您好，这个是我生成随机仿射参数以及变换的代码

def random_affine_transform(image):
    N, C, H, W = image.shape
    scale = round(random.uniform(0.95, 1.15), 1) 
    translate_x = round(random.uniform(-0.05, 0.05), 3)
    translate_y = round(random.uniform(-0.05, 0.05), 3)
    rotate_angle = round(random.uniform(-math.pi/20, math.pi/20), 2)
    # shear_angle = round(random.uniform(-math.pi/6, math.pi/6), 2)    
    cos_angle = math.cos(rotate_angle)
    sin_angle = math.sin(rotate_angle)
    # shear = math.tan(shear_angle)
    theta = torch.tensor([[
        [scale * cos_angle, -scale * sin_angle, translate_x],
        [scale * sin_angle, scale * cos_angle, translate_y]
    ]], dtype=torch.float)
    grid = F.affine_grid(theta, size=image.shape, align_corners=False)
    transformed_image = F.grid_sample(image, grid, align_corners=False)
    return transformed_image, theta

**Comment by Ahuer-Lei** [comment]
![321](https://github.com/Ahuer-Lei/ADRNet/assets/115918220/b8d8883f-aa6f-4616-b00b-fac490817436)
这是我用你代码。生成的扭曲图像和配准图像。
img = torch.tensor(cv2.imread(r"C:\Users\giga\Desktop\1.jpg", cv2.IMREAD_GRAYSCALE)).float().unsqueeze(0).unsqueeze(0)

scale = round(random.uniform(0.95, 1.15), 1)
translate_x = round(random.uniform(-0.05, 0.05), 3)
translate_y = round(random.uniform(-0.05, 0.05), 3)
rotate_angle = round(random.uniform(-math.pi/20, math.pi/20), 2)

cos_angle = math.cos(rotate_angle)
sin_angle = math.sin(rotate_angle)

theta = torch.tensor([[
[scale * cos_angle, -scale * sin_angle, translate_x],
[scale * sin_angle, scale * cos_angle, translate_y]
]], dtype=torch.float)

a = torch.tensor([[[0,0,1]]]).float()
mat = torch.cat((theta, a), dim=1)
mat_inv = torch.inverse(mat)
inv_aff = mat_inv[:, :-1, :]

grid = F.affine_grid(theta, size=[1,1,256,256], align_corners=False)
transformed_image = F.grid_sample(img, grid, align_corners=False)

grid_inv = F.affine_grid(inv_aff, size=[1,1,256,256], align_corners=False)
reg_image = F.grid_sample(transformed_image, grid_inv, align_corners=False)


cv2.imwrite("C:/Users/giga/Desktop/1_1.jpg", transformed_image.squeeze(0).squeeze(0).numpy())
cv2.imwrite("C:/Users/giga/Desktop/1_2.jpg", reg_image.squeeze(0).squeeze(0).numpy())

这是代码， 从我的图片上看是没有大偏移的，

**Comment by huxiaopang666** [comment]
作者您好，我在实验中确实存在这样的情况，您在图中生成的这样的仿射变换moving image相当于原图缩小了，但是还保留了原图里的全部信息，因此逆仿射变换可以完全复原回去，但是因为我研究的内容是图像融合，这种情况下像这些带黑边的moving image是使用不了的，需要像您论文中描述的那样进行裁剪出无黑边的图，以此作为刚性非配准图像对。但是裁剪后的图像必然存在信息不完全匹配，您可以看看采用裁剪后的图像能否完全空间复原

**Comment by Ahuer-Lei** [comment]
![5569](https://github.com/Ahuer-Lei/ADRNet/assets/115918220/a63b3bcc-c743-4f3f-891a-9088ca4e422b)
我试验了一下，裁剪后再配准， 依然没出现你说的问题。
为了方便， 加微信交流吧 18815514485

**Comment by xiaomudsg** [comment]
> ![5569](https://private-user-images.githubusercontent.com/115918220/340080884-a63b3bcc-c743-4f3f-891a-9088ca4e422b.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg1OTIwNzAsIm5iZiI6MTcxODU5MTc3MCwicGF0aCI6Ii8xMTU5MTgyMjAvMzQwMDgwODg0LWE2M2IzYmNjLWM3NDMtNGYzZi04OTFhLTkwODhjYTRlNDIyYi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QwMjM2MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zN2NjNDgxMWM5MDQzOWM1Mzc5MDU1NjM0ODY3NjVhNmU5MWFjYWNkZjUzOTEzNmQwOGI4NTg1NjVhY2E3ZTQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uaAPF_dXM2TLpj9ZaZZ8oJE10GX8HSpVtAO7K44BzPo) 我试验了一下，裁剪后再配准， 依然没出现你说的问题。 为了方便， 加微信交流吧 18815514485

作者您好，我也有相同的疑惑，就是原图片经过放射变换后，会导致部分区域出了尺寸被裁剪掉，这样再进行配准的时候，即使校正过来，那些被裁剪掉的部分也会变成黑边，导致丢失信息，请问您最后是怎么解决这个问题的？我看论文中您最后定量评价的时候是用公共区域四个点的RMSE评估，请问您后续会公开评价指标的代码嘛？真的很需要该代码来验证我的刚性配准效果，学生是配准初学者，对于该问题还望您能解惑。

**Comment by Ahuer-Lei** [comment]
> > ![5569](https://private-user-images.githubusercontent.com/115918220/340080884-a63b3bcc-c743-4f3f-891a-9088ca4e422b.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg1OTIwNzAsIm5iZiI6MTcxODU5MTc3MCwicGF0aCI6Ii8xMTU5MTgyMjAvMzQwMDgwODg0LWE2M2IzYmNjLWM3NDMtNGYzZi04OTFhLTkwODhjYTRlNDIyYi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QwMjM2MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zN2NjNDgxMWM5MDQzOWM1Mzc5MDU1NjM0ODY3NjVhNmU5MWFjYWNkZjUzOTEzNmQwOGI4NTg1NjVhY2E3ZTQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uaAPF_dXM2TLpj9ZaZZ8oJE10GX8HSpVtAO7K44BzPo) 我试验了一下，裁剪后再配准， 依然没出现你说的问题。 为了方便， 加微信交流吧 18815514485
> 
> 作者您好，我也有相同的疑惑，就是原图片经过放射变换后，会导致部分区域出了尺寸被裁剪掉，这样再进行配准的时候，即使校正过来，那些被裁剪掉的部分也会变成黑边，导致丢失信息，请问您最后是怎么解决这个问题的？我看论文中您最后定量评价的时候是用公共区域四个点的RMSE评估，请问您后续会公开评价指标的代码嘛？真的很需要该代码来验证我的刚性配准效果，学生是配准初学者，对于该问题还望您能解惑。

+vx
